File: exercises/day13-Keyboard-First-Refactoring/typescript/src/trip/SleighTripService.ts
import { SleightTrip } from './SleightTrip';
import { SleighTripDAO } from './SleighTripDAO';
import { Elf } from '../elf/Elf';
import { ElfSession } from '../elf/ElfSession';
import { ElfNotLoggedInException } from '../exception/Exceptions';

export class SleighTripService {
    getTripsByUser(targetElf: Elf): SleightTrip[] {
        let sleighTrips: SleightTrip[] = [];
        const loggedElf = this.getLoggedUser();
        let isWorkshopFriend = false;

        if (loggedElf !== null) {
            for (const friend of targetElf.getFriends()) {
                if (friend === loggedElf) {
                    isWorkshopFriend = true;
                    break;
                }
            }
            if (isWorkshopFriend) {
                sleighTrips = SleighTripDAO.findTripsByUser(targetElf);
            }
            return sleighTrips;
        } else {
            throw new ElfNotLoggedInException();
        }
    }

    protected getLoggedUser(): Elf | null {
        return ElfSession.getInstance().getLoggedUser();
    }
}

File: exercises/day13-Keyboard-First-Refactoring/java/src/main/java/sleightrip/trip/SleighTripService.java
package sleightrip.trip;

import java.util.ArrayList;
import java.util.List;

import sleightrip.exception.ElfNotLoggedInException;
import sleightrip.elf.Elf;
import sleightrip.elf.ElfSession;

public class SleighTripService {

    public List<SleightTrip> getTripsByUser(Elf targetElf) throws ElfNotLoggedInException {
        List<SleightTrip> sleighTrips = new ArrayList<>();
        Elf loggedElf = getLoggedUser();
        boolean isWorkshopFriend = false;
        if (loggedElf != null) {
            for (Elf friend : targetElf.getFriends()) {
                if (friend.equals(loggedElf)) {
                    isWorkshopFriend = true;
                    break;
                }
            }
            if (isWorkshopFriend) {
                sleighTrips = SleighTripDAO.findTripsByUser(targetElf);
            }
            return sleighTrips;
        } else {
            throw new ElfNotLoggedInException();
        }
    }

    protected Elf getLoggedUser() {
        return ElfSession.getInstance().getLoggedUser();
    }
}
